design/ratelimit-design-new.md

```diff
@@ -0,0 +1,180 @@
+All [descriptor entries defined by Envoy](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#config-route-v3-ratelimit-action) will be supported except `metadata` and `dynamic_metadata`.
+
+### Contour Configuration
+If using global rate limiting, the details of the external RLS must be configured with Contour.
+First, an `ExtensionService` must be defined for the RLS with the cluster-level details for the RLS itself.
+For example:
+```yaml
+apiVersion: projectcontour.io/v1alpha1
+kind: ExtensionService
+metadata:
+  namespace: projectcontour
+  name: ratelimit
+spec:
+  protocol: h2
+  services:
+    - name: ratelimit
+      port: 8081
+  timeoutPolicy:
+    response: 50ms
+```
+
+Second, a new block will be added to the Contour config file to identify the RLS extension service and provide some additional parameters:
+```yaml
+rateLimitService:
+  # extensionService contains the namespace/name of the RLS extension service.
+  extensionService: projectcontour/ratelimit
+  # domain is passed to the RLS for all rate limit requests.
+  domain: contour
+  # denyOnFailure defines whether to fail open or closed. If true, then if the RLS
+  # cannot be reached or does not return a valid rate limit decision within the
+  # specified timeout, the client will receive a 429 response for their request.
+  denyOnFailure: true
+```
+
+### Rate Limit Flows
+
+- First, Envoy applies local rate limiting to incoming requests.
+- If there are no tokens left in the relevant token bucket(s), a `429 Too Many Requests` response is returned.
+- If there was no local rate limiting defined, or there is a token available, Envoy proceeds to global rate limiting.
+- 1+ descriptors, each of which is an ordered list of key/value pairs, are generated for the request based on the global rate limit configuration.
+- The descriptors are sent to the external RLS via gRPC.
+- The external RLS makes a rate limit decision for the request, based on the descriptors it receives.
+- The external RLS returns a response indicating whether the client request should be rate-limited or not.
+- If the request should be rate limited, a `429 Too Many Requests` response is returned to the client along with the `x-envoy-ratelimited` header.
+- If the request should not be rate limited, it is routed to the appropriate upstream cluster and proceeds as normal.
+
+## Alternatives Considered
+
+### Tight integration with Lyft RLS
+Lyft has open-sourced a [reference external RLS implementation](https://github.com/envoyproxy/ratelimit/).
+We considered tightly integrating with this specific RLS implementation, to provide a more streamlined UX for configuring rate limits.
+A CRD could be used -- either `HTTPProxy` or a stand-alone one -- to define actual rate limits, and a controller could convert the CRDs into configuration in the Lyft RLS format.
```

Comment:
This CRD can still be a generic extension point. Similar to how auth is implemented today in Contour, users would write up rate limits which would turn into CRDs that a controller would need to watch for and convert into specific rate limit configuration depending on the implementation. This would match how we integrate with cert-manager today.

Reply:
Yeah noted on line 164.